### ISSUE TITLE
StageName works differently if hard-coded vs dynamic

### ISSUE THREAD

**Issue by specious** [aws/serverless-application-model#2144]
Defining an `AWS::Serverless::API` resource with a dynamically constructed `StageName` produces a  resource instance as if the StageName is empty.

If the `StageName` value is hard-coded, the result is as expected:

```yaml
Resources:
  API:
    Type: AWS::Serverless::Api
    Properties:
      StageName: dev
      Auth:
        Authorizers: []
``` 

Deployment produces a `AWS::ApiGateway::Stage` resource with logical resource id `APIdevStage`.

If the `StageName` property value is changed to a dynamic reference like `!Ref <parameter>` or `!Join ...` or `!Sub "..."`, then the change set generated has the resource logical id as `APIStage` rather than `API<StageName>Stage`.

For example, making a change to the template like this:

```diff
 Resources:
   API:
     Type: AWS::Serverless::Api
     Properties:
-      StageName: dev
+      StageName: !Join [ "-", ["custom", "stagename"] ]
       Auth:
         Authorizers: []
``` 

Produces a deployment changeset like:

```
Initiating deployment
=====================
Uploading to myapp-backend/70b7c6f9a46ab95003c219b6ead729e9.template  8056 / 8056  (100.00%)

Waiting for changeset to be created..

CloudFormation stack changeset
-------------------------------------------------------------------------------------------------------------
Operation                   LogicalResourceId           ResourceType                Replacement               
-------------------------------------------------------------------------------------------------------------
+ Add                       APIDeployment8310cb78c0     AWS::ApiGateway::Deployme   N/A                       
                                                        nt                                                    
+ Add                       APIStage                    AWS::ApiGateway::Stage      N/A                                                                               
* Modify                    API                         AWS::ApiGateway::RestApi    False                     
- Delete                    APIDeploymentab53ec644d     AWS::ApiGateway::Deployme   N/A                       
                                                        nt                                                    
- Delete                    APIdevStage                 AWS::ApiGateway::Stage      N/A                                                                              
-------------------------------------------------------------------------------------------------------------

Changeset created successfully. arn:aws:cloudformation:us-east-1:999999999999:changeSet/samcli-deploy1630975525/3d58e01a-a563-452c-a5f2-df297c3026d0


Previewing CloudFormation changeset before deployment
======================================================
Deploy this changeset? [y/N]:
```

Tested with various SAM versions, including latest from git:

```
$ sam --version
SAM CLI, version 1.30.0
```

Platform: Arch Linux

**Comment by aahung** [comment]
@specious this is expected behavior. We do this because SAM transformation happens before we can resolve the intrinsic functions, not knowing what is actually the stage name:

https://github.com/aws/serverless-application-model/blob/3da746806e0f090f13d49f3a9a296477b56eaf5c/samtranslator/model/api/api_generator.py#L349-L354

Please open a new issue if you have other questions.
